src/net/tls.cc

```diff
@@ -887,8 +897,14 @@ class tls::reloadable_credentials_base {
             }
         }
         void do_callback(std::exception_ptr ep = {}) {
-            if (_cb && !_files.empty()) {
-                _cb(boost::copy_range<std::unordered_set<sstring>>(_files | boost::adaptors::map_keys), std::move(ep));
+            if (_cb) {
+                shared_ptr<const tls::certificate_credentials> creds = nullptr;
```

Comment:
This creates a new certificate set. Why not add a virtual method to reloadable_credentials_base, `virtual tls::certificate_credentials& as_certificate_credentials() const = 0`, that simply returns `*this`, and change your callback to take a reference instead?
This would also ensure we don't generate extra overhead for callbacks that don't care about them, as well as avoid having to potentially generate an extra exception that was not part of the reload cycle until now (race).

Reply:
Great suggestion. Done.